### DISCUSSION TITLE
Multiple settings pages

### DISCUSSION THREAD

**Discussion by iammathew** [ory/kratos#2248] [Q&A] [answered]
Hey!

I'm currently working on a custom ORY Kratos UI using Next.js SSR and running into the following problem. I would like to have 2 different pages for changing "settings", one related to the profile another to security (changing passwords, etc.)

Currently, the flow looks like this:
1. User enters `/settings` and since no flowId is given gets redirected to `/api/.ory/self-service/settings/browser`
2. User gets redirected by ORY Kratos to the self service settings UI configured incl. flow id `/settings` in my case

But when trying to initiate this flow from a different page like `/security` one will still end up at `/settings`, since this is the UI configured for settings workflows for Kratos.
Is there a way to tell Ory Kratos to send back the user to a different URL when initiating a settings self-service flow?
I tried using `return_to` but that redirects after the flow is completed, not after initialization.

**Comment by vinckr** [comment]
Are you running Ory Kratos in Ory Cloud or self-hosting? 

I am not sure if you can override the settings self-service flow, but [__Hooks__](https://www.ory.sh/docs/kratos/self-service/hooks) might be an option.
Or you could structure your settings UI in a way that you do the same flow in the background but show the user a different UI depending on if they choose "normal" settings or "security" settings.

ps: how would your "ideal", preferred flow look like?

**Reply by iammathew** [reply]
Thanks for the fast reply!

I'm running Ory Kratos self-hosted locally.

Regarding Hooks I'm not sure how they can help me here, since they can't change the initialization flow, right?

My ideal flow is actually quite simple:
Both pages should just initiate a settings flow, but one of them provides an overwrite to the default settings UI to redirect back, similar to `return_to` but instead of when the flow is "finished" it returns to it once the flow is initialized.

My kratos config currently contains this:

```
selfservice:
  default_browser_return_url: http://127.0.0.1:3000
  flows:
    settings:
      ui_url: http://127.0.0.1:3000/settings
      privileged_session_max_age: 15m
```

Initiating a self-service settings flow (using `/api/.ory/self-service/settings/browser`) will always redirect to `ui_url`, so a way of multiple different `ui_url` or a way to dynamically overwrite it (something like `/api/.ory/self-service/settings/browser?ui=http://127.0.0.1:3000/security`)  would be ideal.

This is not identical to `return_to` since this will redirect once you completed the flow aka sent the action.

**Reply by aeneasr** [reply]
Sorry, it looks like this slipped our radar - is this still an issue for you?

**Reply by iammathew** [reply]
@aeneasr Yep, still relevant. I found a workaround by treating the whole auth app as a SPA but that won't work in no JS environments. And in general, it seems like this should also be possible when using the browser flow.

**Reply by aeneasr** [reply]
I see, one way to do it would probably be to initialize the flow with a query parameter of your choosing, for example:

```
/api/.ory/self-service/settings/browser?my_page=something_foo
```

In the settings flow, this query parameter will be visible in the request_url:

https://www.ory.sh/docs/reference/api#operation/getSelfServiceSettingsFlow

```
{
  "request_url": "/api/.ory/self-service/settings/browser?my_page=something_foo"
}
```

you can then use that to display and filter the things you want/need/don't need.

It's probably not the most straight-forward solution, but it will work very reliably and it's extensible! You can for example also add locales like this :) An alternative to that would be to set a HTTP Cookie before you initialize the flow and then once you end up in the UI, decide what to do based on the cookie.

@vinckr this would probably make for a good page on the docs as this question comes up every now and then?

**Reply by vinckr** [reply]
"How to use multiple UIs for browser self-service flows"? 
could be a good [tip](https://www.ory.sh/docs/kratos/debug/troubleshooting) right now, 
I think for a guide we would need a bit more, e.g. the example above. I make a note to create a simple example later on.

**Comment by vinckr** [comment, marked as answer]
quoting [aeneasr ](https://github.com/ory/kratos/discussions/2248#discussioncomment-2546536)
> I see, one way to do it would probably be to initialize the flow with a query parameter of your choosing, for example:
> 
> ```
> /api/.ory/self-service/settings/browser?my_page=something_foo
> ```
> 
> In the settings flow, this query parameter will be visible in the request_url:
> 
> https://www.ory.sh/docs/reference/api#operation/getSelfServiceSettingsFlow
> 
> ```
> {
>   "request_url": "/api/.ory/self-service/settings/browser?my_page=something_foo"
> }
> ```
> 
> you can then use that to display and filter the things you want/need/don't need.
> 
> It's probably not the most straight-forward solution, but it will work very reliably and it's extensible! You can for example also add locales like this :) An alternative to that would be to set a HTTP Cookie before you initialize the flow and then once you end up in the UI, decide what to do based on the cookie.
